Create wallet network switcher

KatunaNorbert opened the pull request:
Fixes #657 .

Changes proposed in this PR:

- Create a new component which enables user to switch the wallet network to the ddo's network

<img width="499" alt="Screen Shot 2021-07-15 at 12 45 25" src="https://user-images.githubusercontent.com/90316/125775761-7f6b6959-6929-475e-9905-5f78e261b5eb.png">

<img width="704" alt="Screen Shot 2021-07-15 at 12 46 04" src="https://user-images.githubusercontent.com/90316/125775816-37cb78c3-2d90-433b-ad94-1d8f6adb33ca.png">

kremalicious commented on src/utils/web3.ts:
```diff
@@ -19,6 +19,46 @@ export interface NetworkObject {
   urlList: string[]
 }
 
+const networks = [
```
during refactor I tried to unify all network metadata getting so ideally we'll get rid of this variable definition, or fill it with data from somewhere else. 

As for the name itself, `getNetworkDisplayName()` is your friend, making sure we control network name from one central place across whole app.

Then I created `useNetworkMetadata()` which exposes metadata about all networks.

And finally I'm still trying to figure out how to get all `ConfigHelperConfig` out of ocean.js into either this network metadata hook, or as part of `useOcean`. Which we then can use to iterate over as those are the networks the switcher should provide. Ultimately we also have to cross-reference it with `appConfig.chainIdsSupported`. So we could already create some loop to iterate over `appConfig.chainIdsSupported` and construct the data with `getOceanConfig()`

 So this then together should have all the metadata we need without hardcoding them here again

KatunaNorbert commented:
I'm trying to use out `addCustomNetwork` method to switch the user wallet network, but It looks like it is not working for default networks

kremalicious commented:
yes good point, there is (currently) no way to programmatically switch MetaMask's built-in networks, but they are all ETH test networks. So would be fine to only expose a switch button for mainnets in the network switcher and for everything else inform user to switch manually. Leaving us only with the switch from a non-ETH network to ETH mainnet. 

I mean, this is Zapper's workaround when I'm on Polygon and click their "Switch to Ethereum" button:

<img width="523" alt="Screen Shot 2021-06-16 at 14 31 04" src="https://user-images.githubusercontent.com/90316/122219205-7dcb3480-ceaf-11eb-9ce0-0faf7431407b.png">

So if we want to show button at all times and assume it's going to work in future, we could go Zapper route and use our modal. But probably simpler to replace button with message for all ETH networks

KatunaNorbert commented:
I added the component to multiple places mention here #657 but not sure what would be the best place to pace it. In the `editComputeDate` or `editAdvancedSettings` might look better inside the form because the action button is over there, and also we don't have the preview component. Also the WalletNetworkSwitcher components needs some design advice.

I added the `Rinkeby` url for the `metadataCacheUri` so I can access edit asset pages, I don't have assets on mainnet

mihaisc commented:
So, first of all the message is wrong. `You are watching OCEAN on ETH Rinkeby but your wallet is connected to ETH` makes no sense in our context. That was just an example from zapper, where you were actually looking at the ocean token. A more accurate message would be `The current asset is on ETH Rinkby but your wallet is connected to ETH`.
![image](https://user-images.githubusercontent.com/26000280/125049491-b377cd80-e0a9-11eb-86a9-ffb980961a82.png)

I am a bit confused about the functionality, shouldn't be this in the web3 feedback component? Something like 
![image](https://user-images.githubusercontent.com/26000280/125050305-88da4480-e0aa-11eb-8669-7eca84842446.png)

KatunaNorbert commented:
Right, makes more sense over there. I'll also fix the message

kremalicious commented on app.config.js:
```diff
@@ -1,7 +1,7 @@
   metadataCacheUri:
     process.env.METADATACACHE_URI ||
-    'https://aquarius.mainnet.oceanprotocol.com',
+    'https://aquarius.rinkeby.oceanprotocol.com',
```
should stay as it was, so fall back to mainnet when env var is not present

kremalicious commented on src/components/atoms/NetworkName.tsx:
```diff
@@ -24,10 +24,12 @@ export function NetworkIcon({ name }: { name: string }): ReactElement {
 export default function NetworkName({
   networkId,
   minimal,
+  textOnly,
```
not sure about the need for this, it might be simpler to access `getNetworkDisplayName()` directly within `WalletNetworkSwither.tsx` if you just want to get the plain text

KatunaNorbert replied:
right, wasn't sure if we need the icon too but if we stick witch the name only the function should do it

kremalicious commented on src/components/molecules/WalletNetworkSwither.tsx:
```diff
@@ -0,0 +1,67 @@
+import React, { ReactElement, useState, useEffect } from 'react'
```
small typo in filename `WalletNetworkSwither.tsx`, missing a `c`: `WalletNetworkSwitcher.tsx`

kremalicious commented on src/components/molecules/WalletNetworkSwither.tsx:
```diff
@@ -0,0 +1,67 @@
+  const { networkId, web3Provider } = useWeb3()
+  const { networksList } = useNetworkMetadata()
+  const { ddo } = useAsset()
+  const DEFAULT_ETH_CHAIN_IDS = [1, 3, 4]
```
they should come from `appConfig.chainIds` which hold the default selected networks

```
const { appConfig } = useSiteMetadata()
```

KatunaNorbert replied:
I think the name is confusing. There I wanted to keep track of the built in Metamask network to know when to show the switch button

kremalicious replied:
ah right, now I get it, thanks! In other news, we might not have to do that anymore, as https://github.com/MetaMask/metamask-extension/pull/11184 seems finally be shipped in latest MetaMask version. I'll test a bit with other dapps and will report back

KatunaNorbert replied:
Oh good news, looking forward for that

kremalicious replied:
so the new Balancer v2 frontend does this where using their network switcher also switches wallet network and it works for ETH Mainnet:

<img width="1359" alt="Screen Shot 2021-07-12 at 16 55 08" src="https://user-images.githubusercontent.com/90316/125309281-eb914180-e331-11eb-8fb5-f01c07115142.png">

And MetaMask docs now have this handy code snippet showing how to combine `wallet_addEthereumChain` with `wallet_switchEthereumChain` so we do not have to maintain a list of builtin networks:

https://docs.metamask.io/guide/rpc-api.html#usage-with-wallet-switchethereumchain

KatunaNorbert replied:
Cool, it's working :)

kremalicious reviewed (approved):
this works like a charm, great job! Think we found all the places where connection to asset network is important but we might find more during testing. Like, was not sure how adding datatoken as custom token behaves across networks with MetaMask so also based this on `isAssetNetwork`.

And for future we need to check other wallets and how network switching behaves with them as we currently show message and action to every wallet but a quick fix then is always to hide the switch button for certain wallets.

KatunaNorbert commented:
Also was thinking to add something to the Publish page, to make it clear for the user to what network is he publishing to, while having multiple networks selected. Also might be publishing to other network than the selected ones if the wallet is connected to different network. But this component doesn't really fit over there because you don't have an asset selected.

kremalicious commented:
yup! Had same idea and tackled this already as part of https://github.com/oceanprotocol/market/pull/729 where the publish page gets this network badge at the top

KatunaNorbert commented:
Oh, nice!
